src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs

```diff
@@ -1277,64 +1277,72 @@ private static unsafe int ApplyMask(Span<byte> toMask, int mask, int maskIndex)
-            if (Vector.IsHardwareAccelerated &&
-                Vector<byte>.Count % sizeof(int) == 0 &&
-                toMask.Length >= Vector<byte>.Count)
-            {
-                Vector<byte> maskVector = Vector.AsVectorByte(new Vector<int>(shiftedMask));
-                Span<Vector<byte>> toMaskVector = MemoryMarshal.Cast<byte, Vector<byte>>(toMask);
-                for (int i = 0; i < toMaskVector.Length; i++)
+            fixed (byte* toMaskBeg = toMask)
+            {
+                byte* toMaskPtr = toMaskBeg;
+                byte* toMaskEnd = toMaskBeg + toMask.Length;
+                byte* maskPtr = (byte*)&mask;
+
+                if (toMaskEnd - toMaskPtr < sizeof(int))
                 {
-                    toMaskVector[i] ^= maskVector;
+                    goto small;
                 }
 
-                // Fall through to processing any remaining bytes that were less than a vector width.
-                toMask = toMask.Slice(Vector<byte>.Count * toMaskVector.Length);
-            }
+                // align our pointer to sizeof(int)
 
-            // If there are any bytes remaining (either we couldn't use vectors, or the count wasn't
-            // an even multiple of the vector width), process them without vectors.
-            int count = toMask.Length;
-            if (count > 0)
-            {
-                fixed (byte* toMaskPtr = &MemoryMarshal.GetReference(toMask))
+                while ((ulong)toMaskPtr % sizeof(int) != 0)
                 {
-                    byte* p = toMaskPtr;
+                    *toMaskPtr++ ^= maskPtr[maskIndex];
+                    maskIndex = (maskIndex + 1) & 3;
+                }
 
-                    // Try to go an int at a time if the remaining data is 4-byte aligned and there's enough remaining.
-                    if (((long)p % sizeof(int)) == 0)
-                    {
-                        while (count >= sizeof(int))
-                        {
-                            count -= sizeof(int);
-                            *((int*)p) ^= shiftedMask;
-                            p += sizeof(int);
-                        }
+                int rolledMask = (int)BitOperations.RotateRight((uint)mask, maskIndex * 8);
+
+                // use SIMD if possible.
+
+                if (Vector.IsHardwareAccelerated && Vector<byte>.Count % sizeof(int) == 0 && (toMaskEnd - toMaskPtr) >= Vector<byte>.Count)
+                {
+                    // align our pointer to Vector<byte>.Count
 
-                        // We don't need to update the maskIndex, as its mod-4 value won't have changed.
-                        // `p` points to the remainder.
+                    while ((ulong)toMaskPtr % (uint)Vector<byte>.Count != 0)
+                    {
+                        *(int*)toMaskPtr ^= rolledMask;
+                        toMaskPtr += sizeof(int);
                     }
```

Comment:
Why do we first align to `sizeof(int)` only to then align to `Vector<byte>.Count`? i.e. why not move the int-alignment to below this whole Vector-block, so that we only do it if toMaskEnd - toMaskPtr >= sizeof(int) after the vectorization?

Reply:
We use `int` alignment it to align the pointer to `Vector`, as you have highlighted here, so we can't move it post-vector.

Reply:
I see.  I missed the `(int*)` cast in the vector alignment loop.